File: Practice/37/C++/37.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

double sqr(double a);
bool equal(double a, double b, double e = 1E-10);

const double PI = 3.141592653589793;
const double precision = 1e-10;

enum type
{
    Cartesian,
    Polar
};

class Point
{
private:
    double x, y;

public:
    Point(double a1 = 0, double a2 = 0, type coord_system = Cartesian)
    {
        if (coord_system == Cartesian)
        {
            x = a1;
            y = a2;
            return;
        }
        else
        {
            x = a1 * cos(a2);
            y = a1 * sin(a2);
        }
    }

    bool operator ==(const Point& second) const
    {
        return abs(x - second.x) < precision && abs(y - second.y) < precision;
    }

    bool operator !=(const Point& second) const
    {
        return !(*this == second);
    }

    double get_x() const
    {
        return x;
    }

    double get_y() const
    {
        return y;
    }

    double get_r() const
    {
        return sqrt(x * x + y * y);
    }

    double get_phi() const
    {
        return atan(y / x) + (x < 0 ? PI : 0);
    }

    void set_x(double x)
    {
        this->x = x;
    }

    void set_y(double y)
    {
        this->y = y;
    }

    void set_r(double r)
    {
        double phi = get_phi();
        this->x = cos(phi) * r;
        this->y = sin(phi) * r;
    }

    void set_phi(double phi)
    {
        double r = get_r();
        this->x = cos(phi) * r;
        this->y = sin(phi) * r;
    }
};

std::ostream& operator <<(std::ostream& out, const Point& point)
{
    return out << '(' << point.get_x() << ',' << point.get_y() << ')';
}

std::istream& operator >>(std::istream& in, Point& point)
{
    double num;
    in.ignore(1);
    in >> num;
    point.set_x(num);
    in.ignore(1);
    in >> num;
    point.set_y(num);
    in.ignore(1);
    return in;
}

class Vector
{
private:
    Point line;

public:
    Vector()
        : line(1, 0)
    {
    }

    Vector(const Point& end)
        : line(end.get_x(), end.get_y())
    {
    }

    Vector(const Point& begin, const Point& end)
        : line(end.get_x() - begin.get_x(), end.get_y() - begin.get_y())
    {
    }

    bool operator ==(const Vector& second) const
    {
        return line == second.line;
    }

    Vector operator -() const
    {
        Point vec = line;
        vec.set_x(-vec.get_x());
        vec.set_y(-vec.get_y());
        return Vector(vec);
    }

    Vector operator +(const Vector& second) const
    {
        Point vec;
        vec.set_x(line.get_x() + second.line.get_x());
        vec.set_y(line.get_y() + second.line.get_y());
        return Vector(vec);
    }

    Vector operator -(const Vector& second) const
    {
        Point vec;
        vec.set_x(line.get_x() - second.line.get_x());
        vec.set_y(line.get_y() - second.line.get_y());
        return Vector(vec);
    }

    Vector operator *(double mul) const
    {
        Point vec = line;
        vec.set_x(vec.get_x() * mul);
        vec.set_y(vec.get_y() * mul);
        return Vector(vec);
    }

    double length() const
    {
        return line.get_r();
    }

    double operator *(const Vector& second) const
    {
        return length() * second.length() * cos(line.get_phi() - second.line.get_phi());
    }
};

int main()
{
    Vector a(Point(1, 2)), b(Point(-2, 0), Point(-1, 2));
    if (a == b && b == a) cout << "Equality test passed\n";
    else cout << "Equality test failed\n";

    Vector na(Point(-1, -2)), ox(Point(1, 0)), nox(Point(-1, 0)), oy(Point(0, 1)), noy(Point(0, -1));
    if (a == -na && na == -a && -ox == nox && -oy == noy) cout << "Invert test passed\n";
    else cout << "Invert test failed\n";

    if (ox + oy + oy == a && -ox == -a + oy + oy) cout << "Summation test passed\n";
    else cout << "Summation test failed\n";

    if (-ox + oy == oy - ox && -oy + ox == ox - oy) cout << "Subtraction test passed\n";
    else cout << "Subtraction test failed\n";

    if (ox * 3 == ox + ox + ox &&
        oy * 3 == oy + oy + oy &&
        ox * (-3) == -ox - ox - ox &&
        oy * (-3) == -oy - oy - oy) cout << "Multiplication by number test passed\n";
    else cout << "Multiplication by number test failed\n";

    if (equal(ox.length(), 1) &&
        equal(oy.length(), 1) &&
        equal((ox * 3 + oy * 4).length(), 5)) cout << "Length test passed\n";
    else cout << "Length test failed\n";

    if (equal(ox * ox, sqr(ox.length())) &&
        equal(oy * oy, sqr(oy.length())) &&
        equal((ox * 3 + oy * 4) * (ox * 3 + oy * 4), sqr((ox * 3 + oy * 4).length()))) cout << "Multiplication by Vector test passed\n";
    else cout << "Multiplication by Vector test failed\n";
}

bool equal(double a, double b, double e) {
    if (-e < a - b && a - b < e) return true;
    else return false;
}

double sqr(double a) {
    return a * a;
}

File: Practice/37/Python/37.py
import copy
import math
import enum

def equal(a, b, e=1E-10):
    if -e < a - b < e: return True
    else: return False

Cartesian = enum.auto()
Polar = enum.auto()

e = 1e-10

class Point:
    def __init__(self, a1 = 0, a2 = 0, coord_system = Cartesian):
        if type(a1) == str:
            p = a1.find(',')
            self.x = float(a1[1 : p].strip())
            self.y = float(a1[p+1 : -2].strip())
            return

        if (coord_system == Cartesian):
            self.x = a1
            self.y = a2
            return
        
        self.x = cos(a2) * a1
        self.y = sin(a2) * a1

    def __eq__(self, second):
        return type(second) == Point and (-e < self.x - second.x < e) and (-e < self.y - second.y < e)

    def __ne__(self, second):
        return not self == second

    def get_x(self):
        return self.x
        
    def set_x(self, x):
        self.x = x

    def get_y(self):
        return self.y
        
    def set_y(self, y):
        self.y = y

    def get_r(self):
        return math.hypot(self.x, self.y)
        
    def set_r(self, r):
        phi = self.get_phi()
        self.x = cos(phi) * r
        self.y = sin(phi) * r

    def get_phi(self):
        return math.atan2(self.y, self.x)

    def set_phi(self, phi):
        r = self.get_r()
        self.x = cos(phi) * r
        self.y = sin(phi) * r

    def __repr__(self):
        return f"Point({self.x},{self.y})"

    def __str__(self):
        return f"({self.x},{self.y})"

class Vector:
    line: Point
    
    def __init__(self, begin = None, end = None):
        if (begin == None) and (end == None):
            self.line = Point(1, 0)
        if (begin != None) and (end == None):
            self.line = begin
        if (begin == None) and (end != None):
            self.line = end
        if (begin != None) and (end != None):
            self.line = Point(end.get_x() - begin.get_x(), end.get_y() - begin.get_y())

    def __eq__(self, second):
        return self.line == second.line

    
    def __neg__(self):
        return Vector(Point(-self.line.get_x(), -self.line.get_y()))

    
    def __add__(self, second):
        vec = Point()
        vec.set_x(self.line.get_x() + second.line.get_x())
        vec.set_y(self.line.get_y() + second.line.get_y())
        
        return Vector(vec)

    
    def __sub__(self, second):
        vec = Point()
        vec.set_x(self.line.get_x() - second.line.get_x())
        vec.set_y(self.line.get_y() - second.line.get_y())
        
        return Vector(vec)

    
    def __mul__(self, second):
        if type(second) == int or type(second) == float:
            vec = Point()
            vec.set_x(self.line.get_x() * second)
            vec.set_y(self.line.get_y() * second)
            
            return Vector(vec)
        else:
            return self.length() * second.length() * math.cos(self.line.get_phi() - second.line.get_phi())

    
    def length(self):
        return self.line.get_r()










#______________________________________________________________________________________________________________

a = Vector(Point(1, 2))
b = Vector(Point(-2, 0), Point(-1, 2))
if a == b and b == a: print('Equality test passed')
else: print('Equality test failed')

na  = Vector(Point(-1, -2))
ox  = Vector(Point( 1,  0))
nox = Vector(Point(-1,  0))
oy  = Vector(Point( 0,  1))
noy = Vector(Point( 0, -1))
if a == -na and na == -a and -ox == nox and -oy == noy: print('Invert test passed')
else: print('Invert test failed')

if ox + oy + oy == a and -ox == -a + oy + oy: print('Summation test passed')
else: print('Summation test failed')

if -ox + oy == oy - ox and -oy + ox == ox - oy: print('Subtraction test passed')
else: print('Subtraction test failed')

if (ox * 3 == ox + ox + ox and
    oy * 3 == oy + oy + oy and
    ox * (-3) == -ox - ox - ox and
    oy * (-3) == -oy - oy - oy): print('Multiplication by number test passed')
else: print('Multiplication by number test failed')

if (equal(ox.length(), 1) and
    equal(oy.length(), 1) and
    equal((ox * 3 + oy * 4).length(), 5)): print('Length test passed')
else: print('Length test failed')

if (equal(ox*ox, ox.length()**2) and
    equal(oy*oy, oy.length()**2) and
    equal((ox*3 + oy*4)*(ox*3 + oy*4), (ox*3 + oy*4).length()**2)): print('Multiplication by Vector test passed')
else: print('Multiplication by Vector test failed')
